File: frontend/src/app/core/services/interview-lab.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import {
  Experience,
  ExperienceCreate,
  ExperienceUpdate,
  FlashCard,
  FlashCardCreate,
  FlashCardUpdate,
  Question,
  QuestionCreate,
  QuestionUpdate,
  Tag,
  TagCreate,
} from '../models/interview-lab.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root',
})
export class InterviewLabService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiBaseUrl}/interview-lab`;

  // Signals for state management
  tags = signal<Tag[]>([]);
  questions = signal<Question[]>([]);
  experiences = signal<Experience[]>([]);
  flashcards = signal<FlashCard[]>([]);

  isLoading = signal<boolean>(false);
  errorMessage = signal<string>('');

  // -------------------------------------------------------------------------
  // Tags API
  // -------------------------------------------------------------------------
  loadTags(): Observable<Tag[]> {
    return this.http.get<Tag[]>(`${this.baseUrl}/tags`).pipe(
      tap((tags) => this.tags.set(tags))
    );
  }

  createTag(payload: TagCreate): Observable<Tag> {
    return this.http.post<Tag>(`${this.baseUrl}/tags`, payload).pipe(
      tap((newTag) => {
        this.tags.update((tags) => {
          const exists = tags.some((t) => t.id === newTag.id || t.name.toLowerCase() === newTag.name.toLowerCase());
          return exists ? tags : [newTag, ...tags];
        });
      })
    );
  }

  deleteTag(tagId: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/tags/${tagId}`).pipe(
      tap(() => {
        this.tags.update((tags) => tags.filter((t) => t.id !== tagId));
      })
    );
  }

  // -------------------------------------------------------------------------
  // Questions API
  // -------------------------------------------------------------------------
  loadQuestions(filter?: {
    applicationId?: string | null;
    company?: string | null;
    difficulty?: string;
    tag?: string;
    search?: string;
  }): Observable<Question[]> {
    this.isLoading.set(true);
    let params = new HttpParams();
    if (filter?.applicationId) {
      params = params.set('application_id', filter.applicationId);
    } else if (filter?.company) {
      params = params.set('company', filter.company);
    }
    if (filter?.difficulty && filter.difficulty !== 'All') {
      params = params.set('difficulty', filter.difficulty);
    }
    if (filter?.tag && filter.tag !== 'All') {
      params = params.set('tag', filter.tag);
    }
    if (filter?.search) {
      params = params.set('search', filter.search);
    }

    return this.http.get<Question[]>(`${this.baseUrl}/questions`, { params }).pipe(
      tap({
        next: (questions) => {
          this.questions.set(questions);
          this.isLoading.set(false);
        },
        error: (err) => {
          this.isLoading.set(false);
          this.errorMessage.set(err?.message || 'Failed to load questions');
        },
      })
    );
  }

  getQuestion(id: string): Observable<Question> {
    return this.http.get<Question>(`${this.baseUrl}/questions/${id}`);
  }

  createQuestion(payload: QuestionCreate): Observable<Question> {
    return this.http.post<Question>(`${this.baseUrl}/questions`, payload).pipe(
      tap((newQ) => {
        this.questions.update((list) => [newQ, ...list]);
      })
    );
  }

  updateQuestion(id: string, payload: QuestionUpdate): Observable<Question> {
    return this.http.put<Question>(`${this.baseUrl}/questions/${id}`, payload).pipe(
      tap((updated) => {
        this.questions.update((list) => list.map((q) => (q.id === id ? updated : q)));
      })
    );
  }

  deleteQuestion(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/questions/${id}`).pipe(
      tap(() => {
        this.questions.update((list) => list.filter((q) => q.id !== id));
      })
    );
  }

  // -------------------------------------------------------------------------
  // Experiences API
  // -------------------------------------------------------------------------
  loadExperiences(filter?: {
    applicationId?: string | null;
    company?: string | null;
    tag?: string;
    minRating?: number;
  }): Observable<Experience[]> {
    this.isLoading.set(true);
    let params = new HttpParams();
    if (filter?.applicationId) {
      params = params.set('application_id', filter.applicationId);
    } else if (filter?.company) {
      params = params.set('company', filter.company);
    }
    if (filter?.tag && filter.tag !== 'All') {
      params = params.set('tag', filter.tag);
    }
    if (filter?.minRating) {
      params = params.set('min_rating', filter.minRating.toString());
    }

    return this.http.get<Experience[]>(`${this.baseUrl}/experiences`, { params }).pipe(
      tap({
        next: (exps) => {
          this.experiences.set(exps);
          this.isLoading.set(false);
        },
        error: (err) => {
          this.isLoading.set(false);
          this.errorMessage.set(err?.message || 'Failed to load interview experiences');
        },
      })
    );
  }

  getExperience(id: string): Observable<Experience> {
    return this.http.get<Experience>(`${this.baseUrl}/experiences/${id}`);
  }

  createExperience(payload: ExperienceCreate): Observable<Experience> {
    return this.http.post<Experience>(`${this.baseUrl}/experiences`, payload).pipe(
      tap((newExp) => {
        this.experiences.update((list) => [newExp, ...list]);
      })
    );
  }

  updateExperience(id: string, payload: ExperienceUpdate): Observable<Experience> {
    return this.http.put<Experience>(`${this.baseUrl}/experiences/${id}`, payload).pipe(
      tap((updated) => {
        this.experiences.update((list) => list.map((e) => (e.id === id ? updated : e)));
      })
    );
  }

  deleteExperience(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/experiences/${id}`).pipe(
      tap(() => {
        this.experiences.update((list) => list.filter((e) => e.id !== id));
      })
    );
  }

  // -------------------------------------------------------------------------
  // Flash Cards API
  // -------------------------------------------------------------------------
  loadFlashcards(filter?: {
    applicationId?: string | null;
    company?: string | null;
    tag?: string;
    difficulty?: string;
  }): Observable<FlashCard[]> {
    this.isLoading.set(true);
    let params = new HttpParams();
    if (filter?.applicationId) {
      params = params.set('application_id', filter.applicationId);
    } else if (filter?.company) {
      params = params.set('company', filter.company);
    }
    if (filter?.tag && filter.tag !== 'All') {
      params = params.set('tag', filter.tag);
    }
    if (filter?.difficulty && filter.difficulty !== 'All') {
      params = params.set('difficulty', filter.difficulty);
    }

    return this.http.get<FlashCard[]>(`${this.baseUrl}/flashcards`, { params }).pipe(
      tap({
        next: (cards) => {
          this.flashcards.set(cards);
          this.isLoading.set(false);
        },
        error: (err) => {
          this.isLoading.set(false);
          this.errorMessage.set(err?.message || 'Failed to load flashcards');
        },
      })
    );
  }

  getStudyDeck(options?: {
    applicationId?: string | null;
    company?: string | null;
    count?: number;
    tag?: string;
    difficulty?: string;
    shuffle?: boolean;
  }): Observable<FlashCard[]> {
    let params = new HttpParams();
    if (options?.applicationId) params = params.set('application_id', options.applicationId);
    else if (options?.company) params = params.set('company', options.company);
    if (options?.count) params = params.set('count', options.count.toString());
    if (options?.tag && options.tag !== 'All') params = params.set('tag', options.tag);
    if (options?.difficulty && options.difficulty !== 'All') params = params.set('difficulty', options.difficulty);
    if (options?.shuffle !== undefined) params = params.set('shuffle', options.shuffle.toString());

    return this.http.get<FlashCard[]>(`${this.baseUrl}/flashcards/study`, { params });
  }

  createFlashcard(payload: FlashCardCreate): Observable<FlashCard> {
    return this.http.post<FlashCard>(`${this.baseUrl}/flashcards`, payload).pipe(
      tap((newCard) => {
        this.flashcards.update((list) => [newCard, ...list]);
      })
    );
  }

  updateFlashcard(id: string, payload: FlashCardUpdate): Observable<FlashCard> {
    return this.http.put<FlashCard>(`${this.baseUrl}/flashcards/${id}`, payload).pipe(
      tap((updated) => {
        this.flashcards.update((list) => list.map((c) => (c.id === id ? updated : c)));
      })
    );
  }

  deleteFlashcard(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/flashcards/${id}`).pipe(
      tap(() => {
        this.flashcards.update((list) => list.filter((c) => c.id !== id));
      })
    );
  }

  // -------------------------------------------------------------------------
  // Excel Export
  // -------------------------------------------------------------------------
  exportExcel(applicationId?: string | null, company?: string | null): void {
    let queryParams = [];
    if (applicationId) queryParams.push(`application_id=${encodeURIComponent(applicationId)}`);
    if (company) queryParams.push(`company=${encodeURIComponent(company)}`);
    const qs = queryParams.length > 0 ? `?${queryParams.join('&')}` : '';
    const downloadUrl = `${this.baseUrl}/export/excel${qs}`;
    window.open(downloadUrl, '_blank');
  }
}

File: services/application/app/services/interview_lab_service.py
"""Business logic and database service layer for Interview Lab."""

from datetime import datetime
import io
import random
from beanie import PydanticObjectId
import openpyxl
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter
import structlog

from app.models.interview_lab import (
    ExperienceDocument,
    FlashCardDocument,
    QuestionDocument,
    TagDocument,
)
from shared.schemas.interview_lab import (
    ExperienceCreate,
    ExperienceResponse,
    ExperienceUpdate,
    FlashCardCreate,
    FlashCardResponse,
    FlashCardUpdate,
    QuestionCreate,
    QuestionResponse,
    QuestionUpdate,
    TagCreate,
    TagResponse,
    TagUpdate,
)

logger = structlog.get_logger()


class InterviewLabService:
    """Service handling CRUD operations and Excel reports for Interview Lab entities."""

    # -------------------------------------------------------------------------
    # Tags
    # -------------------------------------------------------------------------
    @staticmethod
    async def list_tags() -> list[TagResponse]:
        tags = await TagDocument.find_all().sort("-created_at").to_list()
        return [t.to_response_dto() for t in tags]

    @staticmethod
    async def create_tag(payload: TagCreate) -> TagResponse:
        clean_name = payload.name.strip()
        existing = await TagDocument.find_one(TagDocument.name == clean_name)
        if existing:
            return existing.to_response_dto()

        doc = TagDocument(name=clean_name, color=payload.color or "#c25e2e")
        await doc.insert()
        logger.info("interview_tag_created", name=clean_name)
        return doc.to_response_dto()

    @staticmethod
    async def update_tag(tag_id: str, payload: TagUpdate) -> TagResponse | None:
        try:
            oid = PydanticObjectId(tag_id)
        except Exception:
            return None

        doc = await TagDocument.get(oid)
        if not doc:
            return None

        if payload.name is not None:
            doc.name = payload.name.strip()
        if payload.color is not None:
            doc.color = payload.color

        await doc.save()
        return doc.to_response_dto()

    @staticmethod
    async def delete_tag(tag_id: str) -> bool:
        try:
            oid = PydanticObjectId(tag_id)
        except Exception:
            return False

        doc = await TagDocument.get(oid)
        if not doc:
            return False

        await doc.delete()
        logger.info("interview_tag_deleted", tag_id=tag_id)
        return True

    # -------------------------------------------------------------------------
    # Questions
    # -------------------------------------------------------------------------
    @staticmethod
    async def list_questions(
        application_id: str | None = None,
        company: str | None = None,
        difficulty: str | None = None,
        tag: str | None = None,
        search: str | None = None,
        limit: int = 200,
        offset: int = 0,
    ) -> list[QuestionResponse]:
        query = QuestionDocument.find()

        if application_id:
            query = query.find(QuestionDocument.application_id == application_id)
        elif company:
            query = query.find({"company": {"$regex": company, "$options": "i"}})

        if difficulty and difficulty.lower() != "all":
            query = query.find(QuestionDocument.difficulty == difficulty)

        if tag and tag.lower() != "all":
            query = query.find({"tags": tag})

        if search:
            regex_search = {"$regex": search, "$options": "i"}
            query = query.find(
                {
                    "$or": [
                        {"title": regex_search},
                        {"topic": regex_search},
                        {"description": regex_search},
                        {"notes": regex_search},
                    ]
                }
            )

        docs = await query.sort("-updated_at").skip(offset).limit(limit).to_list()
        return [q.to_response_dto() for q in docs]

    @staticmethod
    async def get_question(question_id: str) -> QuestionResponse | None:
        try:
            oid = PydanticObjectId(question_id)
        except Exception:
            return None

        doc = await QuestionDocument.get(oid)
        return doc.to_response_dto() if doc else None

    @staticmethod
    async def create_question(payload: QuestionCreate) -> QuestionResponse:
        doc = QuestionDocument(
            title=payload.title,
            description=payload.description,
            difficulty=payload.difficulty,
            topic=payload.topic,
            application_id=payload.application_id,
            company=payload.company,
            role=payload.role,
            solutions=payload.solutions,
            links=payload.links,
            tags=payload.tags,
            notes=payload.notes,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        await doc.insert()
        logger.info("interview_question_created", title=doc.title)
        return doc.to_response_dto()

    @staticmethod
    async def update_question(question_id: str, payload: QuestionUpdate) -> QuestionResponse | None:
        try:
            oid = PydanticObjectId(question_id)
        except Exception:
            return None

        doc = await QuestionDocument.get(oid)
        if not doc:
            return None

        update_dict = payload.model_dump(exclude_unset=True)
        for key, val in update_dict.items():
            setattr(doc, key, val)

        doc.updated_at = datetime.utcnow()
        await doc.save()
        logger.info("interview_question_updated", question_id=question_id)
        return doc.to_response_dto()

    @staticmethod
    async def delete_question(question_id: str) -> bool:
        try:
            oid = PydanticObjectId(question_id)
        except Exception:
            return False

        doc = await QuestionDocument.get(oid)
        if not doc:
            return False

        await doc.delete()
        logger.info("interview_question_deleted", question_id=question_id)
        return True

    # -------------------------------------------------------------------------
    # Experiences
    # -------------------------------------------------------------------------
    @staticmethod
    async def list_experiences(
        application_id: str | None = None,
        company: str | None = None,
        tag: str | None = None,
        min_rating: int | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[ExperienceResponse]:
        query = ExperienceDocument.find()

        if application_id:
            query = query.find(ExperienceDocument.application_id == application_id)
        elif company:
            query = query.find({"company": {"$regex": company, "$options": "i"}})

        if tag and tag.lower() != "all":
            query = query.find({"tags": tag})

        if min_rating:
            query = query.find(ExperienceDocument.rating >= min_rating)

        docs = await query.sort("-created_at").skip(offset).limit(limit).to_list()
        return [exp.to_response_dto() for exp in docs]

    @staticmethod
    async def get_experience(exp_id: str) -> ExperienceResponse | None:
        try:
            oid = PydanticObjectId(exp_id)
        except Exception:
            return None

        doc = await ExperienceDocument.get(oid)
        return doc.to_response_dto() if doc else None

    @staticmethod
    async def create_experience(payload: ExperienceCreate) -> ExperienceResponse:
        doc = ExperienceDocument(
            company=payload.company,
            role=payload.role,
            application_id=payload.application_id,
            date=payload.date or datetime.utcnow(),
            interview_process=payload.interview_process,
            questions_asked=payload.questions_asked,
            rating=payload.rating,
            overall_notes=payload.overall_notes,
            links=payload.links,
            tags=payload.tags,
            outcome=payload.outcome,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        await doc.insert()
        logger.info("interview_experience_created", company=doc.company, role=doc.role)
        return doc.to_response_dto()

    @staticmethod
    async def update_experience(exp_id: str, payload: ExperienceUpdate) -> ExperienceResponse | None:
        try:
            oid = PydanticObjectId(exp_id)
        except Exception:
            return None

        doc = await ExperienceDocument.get(oid)
        if not doc:
            return None

        update_dict = payload.model_dump(exclude_unset=True)
        for key, val in update_dict.items():
            setattr(doc, key, val)

        doc.updated_at = datetime.utcnow()
        await doc.save()
        logger.info("interview_experience_updated", exp_id=exp_id)
        return doc.to_response_dto()

    @staticmethod
    async def delete_experience(exp_id: str) -> bool:
        try:
            oid = PydanticObjectId(exp_id)
        except Exception:
            return False

        doc = await ExperienceDocument.get(oid)
        if not doc:
            return False

        await doc.delete()
        logger.info("interview_experience_deleted", exp_id=exp_id)
        return True

    # -------------------------------------------------------------------------
    # Flash Cards
    # -------------------------------------------------------------------------
    @staticmethod
    async def list_flashcards(
        application_id: str | None = None,
        company: str | None = None,
        tag: str | None = None,
        difficulty: str | None = None,
        limit: int = 200,
        offset: int = 0,
    ) -> list[FlashCardResponse]:
        query = FlashCardDocument.find()

        if application_id:
            query = query.find(FlashCardDocument.application_id == application_id)
        elif company:
            query = query.find({"company": {"$regex": company, "$options": "i"}})

        if tag and tag.lower() != "all":
            query = query.find({"tags": tag})

        if difficulty and difficulty.lower() != "all":
            query = query.find(FlashCardDocument.difficulty == difficulty)

        docs = await query.sort("-updated_at").skip(offset).limit(limit).to_list()
        return [fc.to_response_dto() for fc in docs]

    @staticmethod
    async def get_flashcard(card_id: str) -> FlashCardResponse | None:
        try:
            oid = PydanticObjectId(card_id)
        except Exception:
            return None

        doc = await FlashCardDocument.get(oid)
        return doc.to_response_dto() if doc else None

    @staticmethod
    async def get_study_deck(
        application_id: str | None = None,
        count: int = 20,
        tag: str | None = None,
        difficulty: str | None = None,
        shuffle: bool = True,
    ) -> list[FlashCardResponse]:
        """Fetch flashcards for active study session."""
        query = FlashCardDocument.find()

        if application_id:
            query = query.find(FlashCardDocument.application_id == application_id)

        if tag and tag.lower() != "all":
            query = query.find({"tags": tag})

        if difficulty and difficulty.lower() != "all":
            query = query.find(FlashCardDocument.difficulty == difficulty)

        docs = await query.to_list()
        if shuffle:
            random.shuffle(docs)

        selected = docs[:count] if count > 0 else docs
        return [fc.to_response_dto() for fc in selected]

    @staticmethod
    async def create_flashcard(payload: FlashCardCreate) -> FlashCardResponse:
        doc = FlashCardDocument(
            front=payload.front,
            back=payload.back,
            application_id=payload.application_id,
            company=payload.company,
            tags=payload.tags,
            links=payload.links,
            difficulty=payload.difficulty,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        await doc.insert()
        logger.info("flashcard_created", front_snippet=doc.front[:30])
        return doc.to_response_dto()

    @staticmethod
    async def update_flashcard(card_id: str, payload: FlashCardUpdate) -> FlashCardResponse | None:
        try:
            oid = PydanticObjectId(card_id)
        except Exception:
            return None

        doc = await FlashCardDocument.get(oid)
        if not doc:
            return None

        update_dict = payload.model_dump(exclude_unset=True)
        for key, val in update_dict.items():
            setattr(doc, key, val)

        doc.updated_at = datetime.utcnow()
        await doc.save()
        logger.info("flashcard_updated", card_id=card_id)
        return doc.to_response_dto()

    @staticmethod
    async def delete_flashcard(card_id: str) -> bool:
        try:
            oid = PydanticObjectId(card_id)
        except Exception:
            return False

        doc = await FlashCardDocument.get(oid)
        if not doc:
            return False

        await doc.delete()
        logger.info("flashcard_deleted", card_id=card_id)
        return True

    # -------------------------------------------------------------------------
    # Excel Export
    # -------------------------------------------------------------------------
    @staticmethod
    async def export_excel(application_id: str | None = None, company: str | None = None) -> bytes:
        """Generate a multi-sheet styled Excel (.xlsx) file containing Experiences, Questions, and Flashcards."""
        exps = await InterviewLabService.list_experiences(application_id=application_id, company=company, limit=500)
        questions = await InterviewLabService.list_questions(application_id=application_id, company=company, limit=500)
        flashcards = await InterviewLabService.list_flashcards(application_id=application_id, company=company, limit=500)

        wb = openpyxl.Workbook()

        # Styles
        header_fill = PatternFill(start_color="1E293B", end_color="1E293B", fill_type="solid")
        header_font = Font(name="Calibri", size=11, bold=True, color="FFFFFF")
        regular_font = Font(name="Calibri", size=10)
        bold_font = Font(name="Calibri", size=10, bold=True)
        thin_border = Border(
            left=Side(style="thin", color="E2E8F0"),
            right=Side(style="thin", color="E2E8F0"),
            top=Side(style="thin", color="E2E8F0"),
            bottom=Side(style="thin", color="E2E8F0"),
        )
        zebra_fill = PatternFill(start_color="F8FAFC", end_color="F8FAFC", fill_type="solid")

        # ---------------------------------------------------------------------
        # Sheet 1: Interview Experiences
        # ---------------------------------------------------------------------
        ws_exp = wb.active
        ws_exp.title = "Interview Experiences"

        exp_headers = [
            "Company",
            "Role",
            "Date",
            "Outcome",
            "Self-Rating (1-10)",
            "Process Pipeline Rounds",
            "Questions Asked & Answers",
            "Retrospective & Notes",
            "Tags",
            "Links",
        ]
        ws_exp.append(exp_headers)
        for col_idx in range(1, len(exp_headers) + 1):
            cell = ws_exp.cell(row=1, column=col_idx)
            cell.fill = header_fill
            cell.font = header_font
            cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

        for row_idx, exp in enumerate(exps, start=2):
            pipeline_str = " ➔ ".join([f"R{s.round_number}: {s.round_type} ({s.description})" if s.description else f"R{s.round_number}: {s.round_type}" for s in exp.interview_process])
            qa_str = "\n".join([f"Q: {qa.question}\nA: {qa.answer} [{qa.category}]" for qa in exp.questions_asked])
            links_str = ", ".join([f"{l.label}: {l.url}" for l in exp.links])
            date_str = exp.date.strftime("%Y-%m-%d") if exp.date else ""

            row_data = [
                exp.company,
                exp.role,
                date_str,
                exp.outcome or "Pending",
                f"{exp.rating}/10",
                pipeline_str,
                qa_str,
                exp.overall_notes,
                ", ".join(exp.tags),
                links_str,
            ]
            ws_exp.append(row_data)

            for col_idx in range(1, len(row_data) + 1):
                cell = ws_exp.cell(row=row_idx, column=col_idx)
                cell.font = regular_font
                cell.border = thin_border
                cell.alignment = Alignment(vertical="top", wrap_text=True)
                if row_idx % 2 == 1:
                    cell.fill = zebra_fill

        # ---------------------------------------------------------------------
        # Sheet 2: DSA & LeetCode Questions
        # ---------------------------------------------------------------------
        ws_q = wb.create_sheet(title="DSA & LeetCode Questions")
        q_headers = [
            "Title",
            "Difficulty",
            "Topic",
            "Problem Summary",
            "Solutions Count",
            "Solution Approaches & Complexities",
            "Key Notes",
            "Tags",
            "Links",
        ]
        ws_q.append(q_headers)
        for col_idx in range(1, len(q_headers) + 1):
            cell = ws_q.cell(row=1, column=col_idx)
            cell.fill = header_fill
            cell.font = header_font
            cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

        for row_idx, q in enumerate(questions, start=2):
            solutions_str = "\n\n".join([
                f"[{sol.label}] ({sol.language})\nTime: {sol.time_complexity} | Space: {sol.space_complexity}\nIntuition: {sol.explanation}\nCode:\n{sol.code}"
                for sol in q.solutions
            ])
            links_str = ", ".join([f"{l.label}: {l.url}" for l in q.links])

            row_data = [
                q.title,
                q.difficulty,
                q.topic,
                q.description,
                len(q.solutions),
                solutions_str,
                q.notes,
                ", ".join(q.tags),
                links_str,
            ]
            ws_q.append(row_data)

            for col_idx in range(1, len(row_data) + 1):
                cell = ws_q.cell(row=row_idx, column=col_idx)
                cell.font = regular_font
                cell.border = thin_border
                cell.alignment = Alignment(vertical="top", wrap_text=True)
                if row_idx % 2 == 1:
                    cell.fill = zebra_fill

        # ---------------------------------------------------------------------
        # Sheet 3: Flashcards
        # ---------------------------------------------------------------------
        ws_fc = wb.create_sheet(title="Revision Flashcards")
        fc_headers = [
            "Front (Question / Prompt)",
            "Back (Answer / Explanation)",
            "Difficulty",
            "Tags",
            "Links",
        ]
        ws_fc.append(fc_headers)
        for col_idx in range(1, len(fc_headers) + 1):
            cell = ws_fc.cell(row=1, column=col_idx)
            cell.fill = header_fill
            cell.font = header_font
            cell.alignment = Alignment(horizontal="center", vertical="center", wrap_text=True)

        for row_idx, fc in enumerate(flashcards, start=2):
            links_str = ", ".join([f"{l.label}: {l.url}" for l in fc.links])
            row_data = [
                fc.front,
                fc.back,
                fc.difficulty,
                ", ".join(fc.tags),
                links_str,
            ]
            ws_fc.append(row_data)

            for col_idx in range(1, len(row_data) + 1):
                cell = ws_fc.cell(row=row_idx, column=col_idx)
                cell.font = regular_font
                cell.border = thin_border
                cell.alignment = Alignment(vertical="top", wrap_text=True)
                if row_idx % 2 == 1:
                    cell.fill = zebra_fill

        # Auto-adjust column widths for all sheets
        for sheet in [ws_exp, ws_q, ws_fc]:
            for col in sheet.columns:
                max_len = 0
                col_letter = get_column_letter(col[0].column)
                for cell in col:
                    val_str = str(cell.value or "")
                    first_line = val_str.split("\n")[0]
                    max_len = max(max_len, len(first_line))
                sheet.column_dimensions[col_letter].width = min(max(max_len + 4, 12), 45)

        stream = io.BytesIO()
        wb.save(stream)
        stream.seek(0)
        return stream.getvalue()
